Mathlib/Combinatorics/SimpleGraph/Clique.lean

```diff
@@ -540,12 +541,71 @@ theorem cliqueSet_map_of_equiv (G : SimpleGraph α) (e : α ≃ β) (n : ℕ) :
 
 end CliqueSet
 
+
+/-! ### Clique number -/
+
+
+section CliqueNumber
+
+variable {α : Type*} {G : SimpleGraph α}
+
+/-- The maximal number of vertices in a graph `G`. -/
+noncomputable def cliqueNum (G : SimpleGraph α) : ℕ := sSup {n | ∃ s, G.IsNClique n s}
+
+/-- A maximum clique in a graph `G` is a clique with the largest possible size. -/
+structure IsMaximumClique (G : SimpleGraph α) (s : Finset α) : Prop where
+  (clique : G.IsClique s)
+  (maximum : ∀ t : Finset α, G.IsClique t → #t ≤ #s)
+
+/-- A maximal clique in a graph `G` is a clique that cannot be extended by adding more vertices. -/
+structure IsMaximalClique (G : SimpleGraph α) (s : Finset α) : Prop where
+  (clique : G.IsClique s)
+  (maximal : ∀ t : Finset α, G.IsClique t → s ⊆ t → t = s)
```

Comment:
Xhat is te use case of thrse predicates.

Reply:
You mean why define maximal cliques? Seemed sensible to include it when defining maximum cliques, but I don't think we need it for our particular application right now (unless @ooovi disagrees).

Reply:
I mean both maximal and maximum cliques

Reply:
Right now @ooovi was working on the AM-GM proof of Mantel which needs this notion. We have also in the past needed the clique number when doing some auxiliary stuff for the hypercube / sensitivity conjecture and I want to clean that up as well.

Actually both of these need the independence number (which requires a notion of independence set in the first place), for which we are preparing another PR that's intended to build on this.

Reply:
> Right now @ooovi was working on the AM-GM proof of Mantel which needs this notion. We have also in the past needed the clique number when doing some auxiliary stuff for the hypercube / sensitivity conjecture and I want to clean that up as well.

How much do you need it? Is it like one lemma or two, or do you a full API about them

Reply:
For my specific proof, I need the dual of `lemma maximumClique_card_eq_cliqueNum (s : Finset α) (sm : G.IsMaximumClique s) : #s = G.cliqueNum`, which is the same statement but about maximum independent sets and the coclique number. The dualization is in a seperate PR, #18608.